sungrow/sungrowdiscovery.cpp

```diff
@@ -0,0 +1,161 @@
+        QTimer::singleShot(3000, this, [this] () {
+            qCDebug(dcSungrow()) << "Discovery: Grace period timer triggered.";
+            finishDiscovery();
+        });
+    });
+}
+
+QList<SungrowDiscovery::SungrowDiscoveryResult> SungrowDiscovery::discoveryResults() const
+{
+    return m_discoveryResults;
+}
+
+void SungrowDiscovery::checkNetworkDevice(const NetworkDeviceInfo &networkDeviceInfo)
+{
+    /* Create a Sungrow connection and try to initialize it.
+    * Only if initialized successfully and all information have been fetched correctly from
+    * the device we can assume this is what we are locking for (ip, port, modbus address, correct registers).
+    */
+
+    qCInfo(dcSungrow()) << "Creating Sungrow Modbus TCP connection for" << networkDeviceInfo.address() << "Port:" << m_port << "Slave Address" << m_modbusAddress;
+    SungrowModbusTcpConnection *connection = new SungrowModbusTcpConnection(networkDeviceInfo.address(), m_port, m_modbusAddress, this);
+    connection->modbusTcpMaster()->setTimeout(5000);
+    connection->modbusTcpMaster()->setNumberOfRetries(0);
+    m_connections.append(connection);
+
+    connect(connection, &SungrowModbusTcpConnection::reachableChanged, this, [=](bool reachable){
+        qCInfo(dcSungrow()) << "Sungrow Modbus TCP Connection reachable changed:" << reachable;
+        if (!reachable) {
+            cleanupConnection(connection);
+            return;
+        }
+        qCDebug(dcSungrow()) << "Connected, proceeding with initialization";
+
+        connect(connection, &SungrowModbusTcpConnection::initializationFinished, this, [=](bool success){
+            if (!success) {
+                qCDebug(dcSungrow()) << "Discovery: Initialization failed on" << networkDeviceInfo.address().toString() << "Continue...";
+                cleanupConnection(connection);
+                return;
+            }
+
+            qCInfo(dcSungrow()) << "Discovery: Initialized successfully" << networkDeviceInfo << connection->serialNumber();
+            qCInfo(dcSungrow()) << "    - Protocol number:" << connection->protocolNumber();
+            qCInfo(dcSungrow()) << "    - Protocol version:" << connection->protocolVersion();
+            qCInfo(dcSungrow()) << "    - ARM software version:" << connection->armSoftwareVersion();
+            qCInfo(dcSungrow()) << "    - DSP software version:" << connection->dspSoftwareVersion();
+
+            if (connection->deviceTypeCode() >= 0xd00 &&  connection->deviceTypeCode() <= 0xeff) {
+                SungrowDiscoveryResult result;
+                result.networkDeviceInfo = networkDeviceInfo;
+                result.serialNumber = connection->serialNumber();
+                result.nominalOutputPower = connection->nominalOutputPower();
+                result.deviceType = connection->deviceTypeCode();
+                m_discoveryResults.append(result);
+            }
+
```

Comment:
Since you are not cleaning up the connection on disconnected, you might consider cleaning it up here too. Since you leaked the discovery object, it would also leak the connection objects here.

Reply:
It will be cleaned up anyway. Either on reachable changed false or on discovery finished